src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h

```diff
@@ -0,0 +1,1777 @@
+
+    bool contains(const std::string &key) const {
+      auto next = *this + 1;
+      if (next == node->iter_end())
+        return get_node_val() <= key;
+
+      return (get_node_val() <= key) && (next->get_node_val() > key);
+    }
+
+    uint16_t get_index() const {
+      return index;
+    }
+
+  private:
+    void set_node_key(omap_inner_key_t _lb) const {
+      static_assert(!is_const);
+      omap_inner_key_le_t lb;
+      lb = _lb;
+      node->get_node_key_ptr()[index] = lb;
+    }
+
+    typename crimson::common::maybe_const_t<char, is_const>::type get_node_key_ptr() const {
+      return reinterpret_cast<
+        typename crimson::common::maybe_const_t<char, is_const>::type>(
+	       node->get_node_key_ptr() + index);
+    }
+
+  };
+  using const_iterator = iter_t<true>;
+  using iterator = iter_t<false>;
+
+  struct delta_inner_t {
+    enum class op_t : uint8_t {
+      INSERT,
+      UPDATE,
+      REMOVE,
+    } op;
+    omap_inner_key_le_t key;
+    std::string val;
+
+    void replay(StringKVInnerNodeLayout &l) {
+      switch (op) {
+      case op_t::INSERT: {
+       l.inner_insert(l.string_lower_bound(val), key, val);
+       break;
+      }
+      case op_t::UPDATE: {
+       auto iter = l.find_string_key(val);
+       assert(iter != l.iter_end());
+       l.inner_update(iter, key);
+       break;
+      }
+      case op_t::REMOVE: {
+       auto iter = l.find_string_key(val);
+       assert(iter != l.iter_end());
+       l.inner_remove(iter);
+       break;
+      }
+      default:
+       assert(0 == "Impossible");
+      }
+    }
+
+    bool operator==(const delta_inner_t &rhs) const {
+      return op == rhs.op &&
+             key == rhs.key &&
+             val == rhs.val;
+    }
+  };
+
+public:
+  class delta_inner_buffer_t {
+    std::vector<delta_inner_t> buffer;
+  public:
+    bool empty() const {
+      return buffer.empty();
+    }
+    void insert(
+      const omap_inner_key_t &key,
+      const std::string val) {
+      omap_inner_key_le_t k;
+      k = key;
+      buffer.push_back(
+       delta_inner_t{
+         delta_inner_t::op_t::INSERT,
+         k,
+         val
+       });
+    }
+    void update(
+      const omap_inner_key_t &key,
+      const std::string &val) {
+      omap_inner_key_le_t k;
+      k = key;
+      buffer.push_back(
+       delta_inner_t{
+         delta_inner_t::op_t::UPDATE,
+         k,
+         val
+       });
+    }
+    void remove(std::string val) {
+      buffer.push_back(
+       delta_inner_t{
+         delta_inner_t::op_t::REMOVE,
+         omap_inner_key_le_t(),
+         val
+       });
+    }
+
+    void replay(StringKVInnerNodeLayout &node) {
+      for (auto &i: buffer) {
+        i.replay(node);
+      }
+    }
+    size_t get_bytes() const {
+      size_t size = 0;
+      for (auto &i: buffer) {
+        size += sizeof(i.op_t) + sizeof(i.key) + i.val.size();
+      }
+      return size;
+    }
+    //copy out
+    void encode(ceph::bufferlist &bl) {
```

Comment:
Implement denc for delta_inner_t, and then do the same here for the vector as a whole (there is already a denc implementation for std::vector<T> if T has a denc implementation).

Reply:
There are example denc implementations in seastore_types.h.